packages/plugin-js-packages/src/lib/utils.ts

```diff
@@ -0,0 +1,99 @@
+import { ExecException, exec } from 'node:child_process';
+import { readFile } from 'node:fs/promises';
+import { join } from 'node:path';
+import { fileExists } from '@code-pushup/utils';
+import {
+  JSPackagesPluginConfig,
+  PackageManagerId,
+  jsPackagesPluginConfigSchema,
+} from './config';
+import { FALLBACK_PACKAGE_MANAGER } from './constants';
+import { packageManagers } from './package-managers';
+
+export async function normalizeConfig(config?: JSPackagesPluginConfig) {
+  const jsPackagesPluginConfig = jsPackagesPluginConfigSchema.parse(
+    config ?? {},
+  );
+
+  const {
+    packageManager = await derivePackageManager(),
+    dependencyGroups: dependencyGroupsCfg = [],
+    checks: checksCfg = [],
+    ...jsPackagesPluginConfigRest
+  } = jsPackagesPluginConfig;
+  const checks = [...new Set(checksCfg)];
+  const depGroups = [...new Set(dependencyGroupsCfg)];
+  const pm = packageManagers[packageManager];
+
+  return {
+    ...jsPackagesPluginConfigRest,
+    packageManager: pm,
+    checks,
+    depGroups,
+  };
+}
+
+export async function derivePackageManagerInPackageJson(
+  currentDir = process.cwd(),
+) {
+  if (await fileExists(join(currentDir, 'package.json'))) {
+    const content = JSON.parse(
+      (await readFile(join('package.json'))).toString(),
+    ) as { packageManager?: string };
+    const { packageManager: packageManagerData = '' } = content;
+
+    const [manager = '', version = ''] = packageManagerData.split('@');
+
+    if (manager === 'npm') {
+      return manager;
+    }
+    if (manager === 'pnpm') {
+      return manager;
+    }
+    if (manager === 'yarn') {
+      const majorVersion = Number(version.split('.')[0]);
+      return majorVersion > 1 ? 'yarn-modern' : 'yarn-classic';
+    }
+  }
+  return false;
+}
+
+export async function derivePackageManager(
+  currentDir = process.cwd(),
+): Promise<PackageManagerId> {
+  const pkgManagerFromPackageJson = await derivePackageManagerInPackageJson(
+    currentDir,
+  );
+  if (pkgManagerFromPackageJson) {
+    return pkgManagerFromPackageJson;
+  }
+
+  // Check for lock files
+  if (await fileExists(join(currentDir, 'package-lock.json'))) {
+    return 'npm';
+  } else if (await fileExists(join(currentDir, 'pnpm-lock.yaml'))) {
+    return 'pnpm';
+  } else if (await fileExists(join(currentDir, 'yarn.lock'))) {
+    const yarnVersion = await new Promise<string>((resolve, reject) => {
+      exec(
+        'yarn -v',
+        (error: ExecException | null, stdout: string, stderr: string) => {
+          if (error) {
+            reject(error);
+          }
+          if (stderr) {
+            reject(stderr);
+          }
+
+          resolve(stdout?.toString().trim().at(0) ?? '');
+        },
+      );
+    });
+
+    if (yarnVersion === '2' || yarnVersion === '3') {
+      return 'yarn-modern';
+    }
+    return 'yarn-classic';
+  }
+  return FALLBACK_PACKAGE_MANAGER;
```

Comment:
maybe it should end with error message hinting user to configure the package manager manually as it could not be determined

Reply:
well if nothing is there then they never installed the packages. or?

Reply:
there could be other things that install deps, they could also just copy them over from cache and never have package manager binary installed.

Also JS ecosystem moves fast, `bun` can install packages as far as I know too 😅

Reply:
what I currently know of:
- npm
- pnpm
- Yarn
- Yarn 2
- Bower
- jspm
- volta
- Entropic
- Bun

Reply:
Thx for the more information!

Just started to introduce an error.

Reply:
done